### PR TITLE
Lib authors options

### LINKED ISSUE #21759: Offer very detailed examples for library authors
_(Directed here from https://github.com/dotnet/AspNetCore.Docs/issues/20799)_

I'm having a difficult time piecing together exactly how library authors should be leveraging the options pattern.

Ideally a whole section of examples where the options pattern is configured/leveraged from within `IServiceCollection` extension methods would really be helpful.

It would be nice to see every technique detailed, including individual and combinations of: 

 - How to leverage strongly typed configuration data while registering services
 - How to allow users to provide their own `Action<OptionsType>` on top of configuration obtained from the options pattern
 - How to establish a default `.Configure` for conventions based section binding
 - Any way to work with `IConfiguration` without requiring users to pass it in to extension methods

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 433988ea-fda9-aa83-f6ee-221a5521a422
* Version Independent ID: 00449cdc-b8fc-a9ff-3f20-0ef0994db354
* Content: [Options pattern in .NET](https://docs.microsoft.com/en-us/dotnet/core/extensions/options)
* Content Source: [docs/core/extensions/options.md](https://github.com/dotnet/docs/blob/master/docs/core/extensions/options.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @IEvangelist
* Microsoft Alias: **dapine**

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

Offer detailed examples for library authors of how to expose the options pattern

Fixes #21759

## Preview

✔️ [Options pattern guidance for .NET library authors](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/options-library-authors?branch=pr-en-us-22433)

### CODE REVIEW

**Comment by IEvangelist** [comment]
Close and reopen, assuming there might have been a transient build issue.

**Review by BillWagner — APPROVED** [review]
This LGTM @IEvangelist 

You can :shipit: when ready.

File: docs/core/extensions/options-library-authors.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+In the preceding code, the `AddMyLibraryService`:
+
+- Extends an instance of <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>
+- Calls <xref:Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.AddOptions%60%601(Microsoft.Extensions.DependencyInjection.IServiceCollection)?displayProperty=nameWithType> with the type parameter of `LibraryOptions`
+- Chains a call to <xref:Microsoft.Extensions.Options.OptionsBuilder%601.Configure%2A>, which specifies the default option values
+
+As the library author, specifying default values is up to you.
+
+> [!NOTE]
```
Comment (davidfowl) [inline docs/core/extensions/options-library-authors.md:45]: I would move this to the next section and just mention that you may choose to read from configuration in the default overloads (see next section for more details).

File: docs/core/extensions/options-library-authors.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+> It is possible to bind configuration to an options instance. However, there is a risk of name collisions - which will cause errors. Additionally, when manually binding in this way, you limit the consumption of your options pattern to read-once. Changes to settings will not be re-bound, as such consumers will not be able to use the [IOptionsMonitor](options.md#ioptionsmonitor) interface.
+>
+> ```csharp
+> services.AddOptions<LibraryOptions>()
```
Comment (davidfowl) [inline docs/core/extensions/options-library-authors.md:49]: I think we should do a better job talking about the various ways users may configure your component once you use Options, and what expectations they might have.

Comment (IEvangelist) [reply docs/core/extensions/options-library-authors.md:49]: How would you phrase this? In what ways might consumers configure the component, are you talking about the various config providers? What expectations would they have, I'm not really sure?

File: docs/core/extensions/options-library-authors.md (hunk 1 above)
Comment (IEvangelist) [reply docs/core/extensions/options-library-authors.md:45]: So rather than having an alert, styling as a callout - mention it in the next section?

Comment (davidfowl) [reply docs/core/extensions/options-library-authors.md:45]: Yep it belongs with the configuration section.

File: docs/core/extensions/options-library-authors.md (hunk 2 above)
Comment (davidfowl) [reply docs/core/extensions/options-library-authors.md:49]: Yes the intro section should describe the benefits of using options in your library and explain what options consumers of the library have configuring it. For example, they can configure options without you providing a specific overload for handling configuration.

**Pull request merged**